jlm/llvm/opt/alias-analyses/PrecisionEvaluator.hpp

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright 2025 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_PRECISIONEVALUATOR_HPP
+#define JLM_LLVM_OPT_ALIAS_ANALYSES_PRECISIONEVALUATOR_HPP
+
+#include <jlm/llvm/opt/alias-analyses/PointsToGraph.hpp>
+#include <jlm/rvsdg/RvsdgModule.hpp>
+#include <jlm/util/Statistics.hpp>
+
+#include <unordered_map>
+
+namespace jlm::llvm::aa
+{
+
+/**
+ * Interface for making alias analysis queries about pairs of pointers
+ */
+class PairwiseAliasAnalysis
+{
+public:
+  PairwiseAliasAnalysis();
+  virtual ~PairwiseAliasAnalysis();
+
+  /**
+   * If this alias analysis is unable to determine NoAlias, it queries the optional backup analysis.
+   * @param backup the PairwiseAliasAnalysis instance to use as backup
+   */
+  void
+  SetBackup(PairwiseAliasAnalysis * backup) noexcept
+  {
+    Backup_ = backup;
+  }
+
+  PairwiseAliasAnalysis *
+  GetBackup() const noexcept
+  {
+    return Backup_;
+  }
+
+  /**
+   * Checks if the pointers represented by p1 and p2 may alias.
+   * If this analysis is unable to prove No Alias, and a backup is provided, the backup is asked.
+   * @param p1 the first pointer value
+   * @param p2 the second pointer value
+   * @return false if the analysis is able to prove that they do not alias, true otherwise
+   */
+  bool
+  MayAlias(const rvsdg::output & p1, const rvsdg::output & p2);
```

Comment:
Would it possible to generalize this where this might be: `AliasInformation GetAliasInformation(...)`, where AliasInformation is:

```
enum AliasInformation
{
   NoAlias,
   MayAlias
}
```

This could then also be extended with potential MustAlias at some point etc.